components/feature/prompts/src/main/java/mozilla/components/feature/prompts/dialog/SaveLoginDialogFragment.kt

```diff
@@ -289,77 +333,38 @@ internal class SaveLoginDialogFragment : PromptDialogFragment() {
     /**
      * Check current state then update view state to match.
      */
-    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
-    fun update() = view?.toScope()?.launch(IO) {
-        val login = Login(
-            guid = guid,
-            origin = origin,
-            formActionOrigin = formActionOrigin,
-            httpRealm = httpRealm,
-            username = username,
-            password = password
-        )
-
-        try {
-            validateStateUpdate?.cancelAndJoin()
-        } catch (cancellationException: CancellationException) {
-            Logger.error("Failed to cancel job", cancellationException)
-        }
-
-        var validateDeferred: Deferred<Result>?
-        validateStateUpdate = launch validate@{
-            val validationDelegate =
-                feature?.loginValidationDelegate ?: return@validate
-            // We only want to fetch this list once. Since we are ignoring username the results will
-            // not change when user changes username or password fields
-            if (potentialDupesList == null) {
-                this@SaveLoginDialogFragment.potentialDupesList = CopyOnWriteArrayList(
-                    validationDelegate.getPotentialDupesIgnoringUsernameAsync(
-                        login
-                    ).await()
-                )
-            }
-            // Passing a copy of this potential dupes list, not a reference, so we know for certain
-            // that what shouldUpdateOrCreateAsync is using can't change.
-            validateDeferred = validationDelegate.shouldUpdateOrCreateAsync(
-                login,
-                potentialDupesList?.toList()
+    private fun update() {
+        // Figure out if we're going to be adding a login or updating an existing one
+        val loginToUpdate = findLoginToUpdate()
+
+        if (loginToUpdate == null) {
+            // Adding a login (or we don't know what's happening because we're
+            // waiting for `loginsStorage.getByBaseDomain()`)
```

Comment:
I don't think I follow. You're not calling getByBaseDomain in this class (unless i missed it). At this point, I would not expect such async behaviour - e.g. you can await for any data you need before update the UI.

Reply:
It's called at the top to setup `loginsForOrigin`.  For me, it seemed slightly simpler to have `update()` not be suspend -- it updates the UI immediately using the currently available data -- and then make sure to call `update()` once the deferred completes.  That way you don't need to cancel the previous job at the top and maybe it simplifies storing the potential dupes list.  That's a very minor preference though.